File: 2020/206.reverse-linked-list.js
/*
 * @lc app=leetcode id=206 lang=javascript
 *
 * [206] Reverse Linked List
 */

class ListNode{
  constructor (val, next) {
     this.val = (val===undefined ? 0 : val)
     this.next = (next===undefined ? null : next)
 }
}
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
  let list = []
  let node = head
  while (node) {
    list.unshift(node.val)
    node = node.next
  }

  head = list.length ? new ListNode(list[0],null) : null
  node = head
  for (let i=1;i<list.length;i++) {
    let swap = new ListNode(list[i],null)
    node.next = swap
    node = node.next
  }
  return head
};
// @lc code=end

File: 206.reverse-linked-list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode], prev=None) -> Optional[ListNode]:
        if head is None: return None

        oldNext = head.next
        head.next = prev
        if oldNext is None:
            return head
        return self.reverseList(oldNext, head)

# @lc code=end

